src/validationinterface.cpp

```diff
@@ -22,6 +27,64 @@ struct CMainSignalsInstance {
     boost::signals2::signal<void (const CBlockIndex *, const std::shared_ptr<const CBlock>&)> NewPoWValidBlock;
 
     CScheduler *scheduler = NULL;
+
+    // We are not allowed to assume the scheduler only runs in one thread,
+    // but must ensure all callbacks happen in-order, so we end up creating
+    // our own queue here :(
+    CCriticalSection cs_callbacksPending;
+    std::list<std::function<void (void)>> callbacksPending;
+    bool fCallbacksRunning = false;
+
+    void MaybeScheduleProcessQueue() {
+        {
+            LOCK(cs_callbacksPending);
+            // Try to avoid scheduling too many copies here, but if we
```

Comment:
@TheBlueMatt Well, in that case I feel like either patch gets funky, especially since the use of fCallbacksRunning becomes inconsistent if you apply the first patch without the second (unless some alternative name for `fCallbacksRunning` works). 

The extra safety check and inconsistency of scheduling bothers me but I wouldn't expect it to actually cause issues so I have no grounds for objection.

I think my inquiry stemmed from it not being immediately apparent that ProcessQueue() only runs once and the extra check is just an extra check. Perhaps the "not a big deal" part of the comment could be "because ProcessQueue() already checks" for clarity, I would not have looked so deeply into the code except that I thought "not a big deal" meant "sometimes interweaving calls is okay."